### DISCUSSION TITLE
Type safety

### DISCUSSION THREAD

**Discussion by kellpossible** [mitsuhiko/minijinja#389] [Ideas]
I've yet to see a templating library for Rust which supports type safety for runtime parsed templates. This would be a killer feature if it were possible to do in minijinja. How could it be done however? Perhaps we can generate a schema for the template at compile time, and ensure that it is stored and associated with the templates which rely on it, and used for type checking when compiling the templates?

**Comment by kellpossible** [comment]
The use case for this for me personally is to have user configurable/editable templates that are still type checked without requiring them to recompile the entire software.

**Comment by mitsuhiko** [comment]
Jinja is intentionally quite dynamically typed. I guess the complexity of type safety would approach the one of type script.

**Reply by kellpossible** [reply]
On this topic https://github.com/rdbo/dynja but it would be nicer to see something that is officially supported with better coverage of functionality.

**Reply by kellpossible** [reply]
Of course that doesn't support runtime type checking. I guess most of the other Rust projects which have static type checking in the templates compile down to Rust and piggy back on the Rust compiler work. Embedding the rust compiler in an application is not desirable, but perhaps something like dynja could be a happy medium.

**Comment by uros-5** [comment]
I created  [language server](https://github.com/uros-5/jinja-lsp) for minijinja last year, it's still not 100% type safe. It can catch some of bugs with warnings.
